File: backend/app/schemas/summary.py
"""
Pydantic v2 schemas for the document summary API (Phase 14).

Request/response models for:
  - GET  /summaries/{document_id}?version=   — get-or-create-and-poll
  - POST /summaries/{document_id}/regenerate — fresh-job regeneration

All response models use from_attributes = True (ORM mode), following
schemas/comparison.py's exact convention (snake_case field names — the FE
client mirrors these; no separate casing convention is introduced).

See PHASE-14-IMPLEMENTATION-PLAN.md §5.13, §13.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class _OrmBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# ── Requests ──────────────────────────────────────────────────────────────────

class SummaryRegenerateRequest(BaseModel):
    """Request body for POST /summaries/{document_id}/regenerate."""

    version: Optional[int] = Field(
        default=None,
        description="Version number to regenerate; null = the document's current version.",
    )


# ── Responses ─────────────────────────────────────────────────────────────────

class SummaryCitation(BaseModel):
    """One inline resolved citation attached to a summary item."""

    index: int
    chunk_id: str
    document_id: str
    document_version_id: str
    document_name: str
    page_id: str
    page_number: int
    section: Optional[str] = None
    relevance: float = 0.0
    quoted_text: str = ""
    char_start: Optional[int] = None
    char_end: Optional[int] = None


class SummaryItem(BaseModel):
    """One summary list item: clean display text + its resolved citations."""

    text: str
    citations: list[SummaryCitation] = Field(default_factory=list)


class SamplingDisclosureSchema(BaseModel):
    """The persisted sampling disclosure (roadmap exit criterion 7 —
    sampling is disclosed, never silent)."""

    sampled: bool = False
    strategy: str = "full"
    included_section_ids: list[str] = Field(default_factory=list)
    excluded_section_count: int = 0


class SummaryPayload(BaseModel):
    """The structured summary content (all fields explicit — an empty
    field is an explicit empty list, never an omitted key)."""

    executive_summary: list[SummaryItem] = Field(default_factory=list)
    key_points: list[SummaryItem] = Field(default_factory=list)
    dates: list[SummaryItem] = Field(default_factory=list)
    roles: list[SummaryItem] = Field(default_factory=list)
    requirements: list[SummaryItem] = Field(default_factory=list)
    risks: list[SummaryItem] = Field(default_factory=list)
    topics: list[str] = Field(default_factory=list)


class SummaryResponse(_OrmBase):
    """Summary status + content (poll target for GET and POST regenerate)."""

    id: str
    document_id: str
    document_version_id: str
    status: str
    summary: Optional[SummaryPayload] = None
    sampling: Optional[SamplingDisclosureSchema] = None
    model: Optional[str] = None
    prompt_version: Optional[str] = None
    stale: bool = False
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None

File: frontend/src/lib/api/summary.ts
/**
 * Typed API functions for document summaries (Phase 14).
 *
 * Matches backend/app/schemas/summary.py and backend/app/api/summaries.py.
 *
 * Endpoints consumed:
 *   GET  /summaries/{documentId}?version=  → getSummary()
 *   POST /summaries/{documentId}/regenerate → regenerateSummary()
 */

import { get, post } from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export type SummaryStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED'

/** One inline resolved citation attached to a summary item. */
export interface SummaryCitation {
  index: number
  chunk_id: string
  document_id: string
  document_version_id: string
  document_name: string
  page_id: string
  page_number: number
  section: string | null
  relevance: number
  quoted_text: string
  char_start: number | null
  char_end: number | null
}

/** One summary list item: clean display text + its resolved citations. */
export interface SummaryItem {
  text: string
  citations: SummaryCitation[]
}

/** The persisted sampling disclosure — sampling is disclosed, never silent. */
export interface SamplingDisclosure {
  sampled: boolean
  strategy: 'full' | 'section_diverse'
  included_section_ids: string[]
  excluded_section_count: number
}

/** The structured summary payload (all fields explicit). */
export interface SummaryPayload {
  executive_summary: SummaryItem[]
  key_points: SummaryItem[]
  dates: SummaryItem[]
  roles: SummaryItem[]
  requirements: SummaryItem[]
  risks: SummaryItem[]
  topics: string[]
}

/** Summary row — from GET /summaries/{documentId} or POST regenerate. */
export interface SummaryResponse {
  id: string
  document_id: string
  document_version_id: string
  status: SummaryStatus
  summary: SummaryPayload | null
  sampling: SamplingDisclosure | null
  model: string | null
  prompt_version: string | null
  stale: boolean
  error_message: string | null
  created_at: string
  completed_at: string | null
}

// ─── API functions ─────────────────────────────────────────────────────────────

/**
 * Get-or-create-and-poll in one endpoint: returns the current summary for
 * the document's current (or explicitly requested) version, creating and
 * enqueuing a SUMMARY job on first access.  Poll until
 * `status === 'COMPLETED' | 'FAILED'`.
 */
export async function getSummary(
  documentId: string,
  version?: number,
): Promise<SummaryResponse> {
  const query = new URLSearchParams()
  if (version != null) query.set('version', String(version))
  const qs = query.toString()
  return get<SummaryResponse>(`/summaries/${documentId}${qs ? `?${qs}` : ''}`)
}

/**
 * Regenerate the summary — always a fresh job; the row resets to PENDING
 * while the existing content stays visible.  Requires `summary:regenerate`.
 */
export async function regenerateSummary(
  documentId: string,
  version?: number,
): Promise<SummaryResponse> {
  return post<SummaryResponse>(`/summaries/${documentId}/regenerate`, {
    version: version ?? null,
  })
}
